### PR TITLE
Add ability to shard customer tests

### PR DESCRIPTION
**Pull request by CaseyHillers**
## Description

Adds the options `--shards int` and `--shard-index int` to `customer_testing/run_tests.dart`. This allows for sharding these tests to reduce the time they take to run.

This is intended for only running on flutter/tests repo cirrus, however, I added it to this repo so it can be used in the future for sharding on the flutter/flutter repo if needed.

## Related Issues

https://github.com/flutter/tests/pull/26#discussion_r397453538 (comment suggesting we shard flutter/tests instead of increasing timeout)

## Tests

I didn't see any tests for this script, so I manually verified different edge cases:

(1) All tests run if no sharding info is provided
(2) All tests have their own shard if number of tests == number of shards
(3) All tests are assigned a shard if there is an unequal number of tests per shard

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: dev/customer_testing/run_tests.dart [hunk 1]
```diff
@@ -25,6 +25,18 @@ Future<bool> run(List<String> arguments) async {
...
+    ..addOption(
+      'shards',
+      defaultsTo: '1',
+      help: 'How many shards to split the tests into. Used in continuous integration.',
+      valueHelp: 'count',
+    )
+    ..addOption(
+      'shard-index',
+      defaultsTo: '0',
+      help: 'The current shard to run the tests as. Used in continuous integration.',
```
Comment (Hixie) [inline dev/customer_testing/run_tests.dart]: it should say what the valid range is (presumably 0..shard-1 ?)

File: dev/customer_testing/run_tests.dart [hunk 2]
```diff
@@ -91,6 +105,21 @@ Future<bool> run(List<String> arguments) async {
...
+  }
+
+  if (verbose) {
+    print('Tests in this shard:');
+    for (final File file in shardedFiles)
+      print(file.path);
+  }
+
   if (verbose)
     print('Starting run_tests.dart...');
```
Comment (Hixie) [inline dev/customer_testing/run_tests.dart]: this should be above all the prints you added, it's supposed to be the first thing in the output. :-)

File: dev/customer_testing/run_tests.dart [hunk 3]
```diff
@@ -99,11 +128,12 @@ Future<bool> run(List<String> arguments) async {
 
   if (verbose) {
     final String s = files.length == 1 ? '' : 's';
-    print('${files.length} file$s specified.');
+    final String ss = shardedFiles.length == 1 ? '' : 's';
+    print('${files.length} file$s specified. ${shardedFiles.length} test$ss in shard #$shardIndex');
```
Comment (Hixie) [inline dev/customer_testing/run_tests.dart]: missing period at the end

**Comment by Hixie** [comment]
> I didn't see any tests for this script, so I manually verified different edge cases:

If you don't see tests, you should add some. :-)

**Comment by Hixie** [comment]
LGTM

File: dev/customer_testing/run_tests.dart (hunk 1 above)
Comment (CaseyHillers) [reply dev/customer_testing/run_tests.dart]: Added. I refrained initially as all the other help lines were short :)

File: dev/customer_testing/run_tests.dart (hunk 2 above)
Comment (CaseyHillers) [reply dev/customer_testing/run_tests.dart]: Thanks for pointing this out. Rearranged the prints to make more sense:

```
chillers@chillzone:~/Documents/flutter$ dart dev/customer_testing/run_tests.dart --verbose --shards 2 --shard-index 0 ~/flutter_tests/registry/*
Starting run_tests.dart...
8 files specified. 4 tests in shard #0.

Tests in this shard:
/usr/local/google/home/chillers/flutter_tests/registry/DanTup_tiler.test
/usr/local/google/home/chillers/flutter_tests/registry/flutter_gallery.test
/usr/local/google/home/chillers/flutter_tests/registry/flutter_svg.test
/usr/local/google/home/chillers/flutter_tests/registry/rainbowmonkey.test

Processing /usr/local/google/home/chillers/flutter_tests/registry/DanTup_tiler.test...
```

File: dev/customer_testing/run_tests.dart (hunk 3 above)
Comment (CaseyHillers) [reply dev/customer_testing/run_tests.dart]: Fixed

**Comment by CaseyHillers** [comment]
Thanks for the review @Hixie, i'll follow up with adding tests to this script in a separate PR since that seems to be a bigger PR.to unblock adding Cocoon to flutter/tests. I figure we can test this script in the same customer testing shard where we can test the script before running it.

**Review by godofredoc — APPROVED** [review]
